File: elixir/dnd-character/lib/dnd_character.ex
defmodule DndCharacter do
  @type t :: %__MODULE__{
          strength: pos_integer(),
          dexterity: pos_integer(),
          constitution: pos_integer(),
          intelligence: pos_integer(),
          wisdom: pos_integer(),
          charisma: pos_integer(),
          hitpoints: pos_integer()
        }

  defstruct ~w[
    strength
    dexterity
    constitution
    intelligence
    wisdom
    charisma
    hitpoints
  ]a

  @die_faces 1..6
  @rolls 0..3
  @initial_hitpoints 10

  @spec modifier(pos_integer()) :: integer()
  def modifier(score), do: Integer.floor_div(score - 10, 2)

  @spec ability :: pos_integer()
  def ability do
    rolls = for _roll <- @rolls, do: Enum.random(@die_faces)
    Enum.sum(rolls) - Enum.min(rolls)
  end

  @spec character :: t()
  def character do
    constitution = ability()

    %DndCharacter{
      strength: ability(),
      dexterity: ability(),
      constitution: constitution,
      intelligence: ability(),
      wisdom: ability(),
      charisma: ability(),
      hitpoints: @initial_hitpoints + modifier(constitution)
    }
  end
end

File: javascript/dnd-character/dnd-character.js
const MIN_SCORE = 3
const MAX_SCORE = 18

export const abilityModifier = score => {
  if (score < MIN_SCORE) {
    throw new Error("Ability scores must be at least 3")
  }

  if (score > MAX_SCORE) {
    throw new Error("Ability scores can be at most 18")
  }

  return Math.floor((score - 10) / 2)
}

export class Character {
  static get MIN_DIE_FACE() {
    return 1
  }

  static get MAX_DIE_FACE() {
    return 6
  }

  static get ROLLS() {
    return [...Array(4).keys()]
  }

  static get INITIAL_HITPOINTS() {
    return 10
  }

  static rollAbility() {
    const rolls = Character.ROLLS.map(Character.roll)
    return Character.sum(rolls) - Math.min(...rolls)
  }

  static sum(rolls) {
    return rolls.reduce((acc, roll) => acc + roll, 0)
  }

  static roll() {
    return (
      Math.floor(
        Math.random() * (Character.MAX_DIE_FACE - Character.MIN_DIE_FACE + 1)
      ) + Character.MIN_DIE_FACE
    )
  }

  constructor() {
    this._strength = Character.rollAbility()
    this._dexterity = Character.rollAbility()
    this._constitution = Character.rollAbility()
    this._intelligence = Character.rollAbility()
    this._wisdom = Character.rollAbility()
    this._charisma = Character.rollAbility()
    this._hitpoints =
      Character.INITIAL_HITPOINTS + abilityModifier(this.constitution)
  }

  get strength() {
    return this._strength
  }

  get dexterity() {
    return this._dexterity
  }

  get constitution() {
    return this._constitution
  }

  get intelligence() {
    return this._intelligence
  }

  get wisdom() {
    return this._wisdom
  }

  get charisma() {
    return this._charisma
  }

  get hitpoints() {
    return this._hitpoints
  }
}
